### ISSUE TITLE
Belarusian and most likely Ukrainian apostrophe is causing troubles for dictionary lookups

### ISSUE THREAD

**Issue by ssvb** [koreader/koreader#10351]
* KOReader version: 2023.03
* Device: gta4xlwifieea (Android Galaxy Tab S6 Lite)
* OS: Android S, api 33 on arm64

#### Issue

Some Belarusian words, such as "надвор'е" or "сур'ёзна", include apostrophe as a part of the word. Some Ukrainian words, such as "пам'ятати", also use it.

#### Steps to reproduce

* Open a epub book and try to tap on any word with an apostrophe. Example: https://knihi.com/Janka_Maur/Paleskija_rabinzony.epub
* Observe that only the part before the apostrophe or the part after the apostrophe gets selected, but not the whole word.

**Comment by hius07** [comment]
Official Ukrainian apostrophe is https://www.codetable.net/decimal/8217

**Comment by ssvb** [comment]
The Belarusian official apostrophe is also U+2019. But in practice ASCII 39 is used a lot and [various spellcheckers](https://bnkorpus.info/spell.en.html) don't complain. I copy-pasted the "пам'ятати" example from the [Ukrainian online dictionary](http://sum.in.ua/s/pam.jataty). Historically, this probably originates from the old 8-bit CP1251 encoding, which used the ascii apostrophe before unicode took over.  Nevertheless the dictionary lookup in `koreader` doesn't work nice with the real Belarusian and Ukrainian epub books right now. A free epub fragment from https://epub.com.ua/books/temna-vezha-temna-vezha-vii/ can be used to reproduce troubles with the Ukrainian words "пов'язані" and "вб'ють".

**Comment by poire-z** [comment]
Same with French "aujourd'hui" ("today"), we get selected "aujourd". It's the engine that gives us the word that is where our finger does long-press - and for the engine, any apostrophe is a word separator (as it would be in the French "d'après" ("from after"). And with "d'aujourd'hui" ("from today") where one apostrophe is an article and should be ignored - and the other should be part of the word).

Anyway, it would need the engine to use the dictionary (or some language support stuff similar to japanese) to extend the selection correctly.
It may have been really necessary for Japanese - but for most other languages, the user can help himself with long-pressing and continue selecting the part that was not selected.

**Comment by ssvb** [comment]
> the user can help himself with long-pressing and continue selecting the part that was not selected

Thanks for this tip! It's better than nothing.

**Comment by ssvb** [comment]
@hius07 Actually everything is a bit more complicated and there are [three apostrophes](https://en.wikipedia.org/wiki/Apostrophe#Unicode) in use. In a strictly pedantic way, the correct Unicode apostrophe for Belarusian and Ukrainian texts is U+02BC (letter apostrophe). This requirement is actually enforced by ICANN when registering Cyrillic domain names, such as http://мʼясо.укр (try to type it in your browser address bar using your keyboard). There's a long discussion thread related to keyboard layout in Linux, where many aspects of this problem are mentioned: https://bugs.freedesktop.org/show_bug.cgi?id=50895

But in the real world not everything is so nice. The keyboard layouts, search engines, fulltext search feature in browsers and text editors, spellcheckers and the other software may have poor support for U+02BC apostrophe. It may be necessary to submit bugreports or patches to Firefox, Chromium, OpenOffice, LanguageTool and the others. I'll try to do some of this as time permits.

Now regarding koreader:
1. The U+02BC apostrophe encountered in ebooks is treated as a part of the word for dictionary lookups via long press. That's very nice! But it's still necessary to ensure that the selected word can be found in the dictionary (without fuzzy search) and there are some problems with that. Many dictionaries seem to be using ASCII apostrophes. Should U+02BC apostrophe be treated in the same way as U+2019 in https://github.com/koreader/koreader/blob/v2023.04/frontend/apps/reader/modules/readerdictionary.lua#L661-L663? Or a better solution is needed?
2. The ASCII and U+2019 apostrophes are not going away in the foreseeable future and they will be encountered in ebooks. So a language plugin to automatically expand the selection area to cover the whole word would be still useful.

**Comment by hius07** [comment]
> Also a very interesting testcase is an Ukrainian translation of Leo Tolstoy's "War and Peace".

Unfortunately the available text of it (at Flibusta) uses ASCII 39 for Ukrainian.

**Comment by ssvb** [comment]
> Unfortunately the available text of it (at Flibusta) uses ASCII 39 for Ukrainian.

Yes, and this is the current reality. Some of the possible ways to deal with it are:
* blame the book publisher and do nothing on the koreader side
* use some external tool to fixup apostrophes in a epub file and then open the conversion result in koreader
* handle different apostrophes in a koreader's languagesupport plugin
* maybe do something else?

**Comment by ssvb** [comment]
> Many dictionaries seem to be using ASCII apostrophes.

Out of ~245K Belarusian lemmas/headwords available in the https://github.com/Belarus/GrammarDB database, only ~1.5K contain apostrophes. So I think that just adding all the possible apostrophe variants as synonyms in a StarDict dictionary will do the job for now and will work nicely with any books. But the user experience of selecting the whole word via long press still can be improved.
